ldap_auth_provider.py

```diff
@@ -245,6 +264,14 @@ def check_3pid_auth(self, medium, address, password):
                 self.ldap_attributes["uid"], [None]
             )
             localpart = localpart[0] if len(localpart) == 1 else None
+            if self.ldap_ad_forest and localpart:
+                [login, domain] = localpart.rsplit("@", 1)
+                localpart = login + self.ldap_separator + domain
+
+                if self.ldap_default_domain and domain == self.ldap_default_domain:
+                    user_id = self.account_handler.get_qualified_user_id(localpart)
+                    if (yield self.account_handler.check_user_exists(user_id)):
+                        localpart = login
```

Comment:
why is the default domain special-cased here? why is it correct to strip the domain for accounts on that domain but not others?

Reply:
2 thoughts here:

1. This is very common pattern in corporate environment.  
You can look eg to [smb.conf's winbind use default domain](https://www.samba.org/samba/docs/current/man-html/smb.conf.5.html#idm10651).
Yes, it may be not recommended, bu it is widelly used.

2. Upgrading from previous versions of ```ldap_auth_provider``` while using ```samAccountName``` as uid.
After this PR merged it will be possible to continue using ```samAccountName``` as matrix uid for default domain users but also allow to use ```userPrincipalName``` for users of other domains in forest.

P.S.: ```samAccountName``` is first part of ```userPrincipalName``` before ```@```